### PR TITLE
CSHARP-2515: Test Planned Maintenance Scenarios in Drivers

### PR DESCRIPTION
**Pull request by MikalaiMazurenka**
JIRA: https://jira.mongodb.org/browse/CSHARP-2515
Evergreen: https://evergreen.mongodb.com/version/5f5689e2c9ec4454b69dc4ea
Native PR: https://github.com/MikalaiMazurenka/mongo-csharp-driver/pull/56

### CODE REVIEW

File: integrations/dotnet/workload-executor [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+export PLATFORM
+
+# Invoke the workload executor as a background process and store its process ID as NATIVE_WORKLOAD_EXECUTOR_PID
+#     We don't need to build or restore because install-driver.sh handles that
+#     We can't use dotnet run --no-build --no-restore due to https://github.com/dotnet/runtime/issues/38274
+#     Addditionally, we must use the published version to bypass a dependency resolution issue with .NET Core 1.1
+set +x # hide connection string in test output
+if [[ "$FRAMEWORK" == "net452" ]]; then
+    ./mongo-csharp-driver/tests/AstrolabeWorkloadExecutor/bin/Debug/"${FRAMEWORK}"/publish/workload-executor.exe \
+        "${CONNECTION_STRING}" "${WORKLOAD_SPEC}"
```
Comment (MikalaiMazurenka) [inline integrations/dotnet/workload-executor:58]: https://github.com/mongodb-labs/drivers-atlas-testing/pull/78/files#r466096338
> There is no need for running this as background processes so we can remove the trailing &.

Done.

File: integrations/dotnet/workload-executor [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+else
+    dotnet mongo-csharp-driver/tests/AstrolabeWorkloadExecutor/bin/Debug/"${FRAMEWORK}"/publish/workload-executor.dll \
+        "${CONNECTION_STRING}" "${WORKLOAD_SPEC}"
```
Comment (MikalaiMazurenka) [inline integrations/dotnet/workload-executor:61]: https://github.com/mongodb-labs/drivers-atlas-testing/pull/78/files#r466096479
> Same comment as above about running this as a background process.

Done.

File: integrations/dotnet/workload-executor [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+fi
+set -x # show output
+export NATIVE_WORKLOAD_EXECUTOR_PID=$!
```
Comment (MikalaiMazurenka) [inline integrations/dotnet/workload-executor:64]: https://github.com/mongodb-labs/drivers-atlas-testing/pull/78/files#r466096592
> We don't need this wait once the executor is run as a foreground process.

Done.

File: .evergreen/config.yml [hunk 4]
```diff
@@ -194,6 +194,19 @@ axes:
           DRIVER_REPOSITORY: "https://github.com/mongodb/mongo-java-driver"
           DRIVER_REVISION: "master"
           ASTROLABE_EXECUTOR_STARTUP_TIME: 15
+      - id: dotnet-csharp2515
```
Comment (prashantmital) [inline .evergreen/config.yml:197]: This should be changed to `dotnet-master`

File: integrations/dotnet/workload-executor [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+echo workload-executor pid: $BASHPID
+
+if [[ "$OS" =~ Windows|windows ]]; then
+  RESULTS_DIR=$(cygpath -w "$(pwd)")
+else
+  RESULTS_DIR="$(pwd)"
+fi
+
+# Set a trap for the INT
```
Comment (prashantmital) [inline integrations/dotnet/workload-executor:40]: Let's remove these comments - I don't think they are relevant anymore.

File: integrations/dotnet/workload-executor [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+#     Addditionally, we must use the published version to bypass a dependency resolution issue with .NET Core 1.1
+set +x # hide connection string in test output
+if [[ "$FRAMEWORK" == "net452" ]]; then
+    ./mongo-csharp-driver/tests/AstrolabeWorkloadExecutor/bin/Debug/"${FRAMEWORK}"/publish/workload-executor.exe \
+        "${CONNECTION_STRING}" "${WORKLOAD_SPEC}"
+else
+    dotnet mongo-csharp-driver/tests/AstrolabeWorkloadExecutor/bin/Debug/"${FRAMEWORK}"/publish/workload-executor.dll \
+        "${CONNECTION_STRING}" "${WORKLOAD_SPEC}"
+fi
+set -x # show output
```
Comment (prashantmital) [inline integrations/dotnet/workload-executor:63]: What does this do?

File: integrations/dotnet/workload-executor (hunk 3 above)
Comment (prashantmital) [inline integrations/dotnet/workload-executor:64]: I think this can be removed.

File: integrations/dotnet/workload-executor [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+#                                   "netcoreapp2.1"
+#   ASYNC                           Whether or not to use async operations (which may be run synchronously). Values: true, false
+# Environment variables produced as output:
+#   RESULTS_DIR                     Where to output results.json
+#   ASYNC                           Whether or not to use async operations (which may be run synchronously). Values: true, false
+
+FRAMEWORK=${FRAMEWORK:-netcoreapp2.1}
+MAGIC_FILE_NAME=${MAGIC_FILE_NAME:-nox}
+ASYNC=${ASYNC:-true}
+############################################
```
Comment (prashantmital) [inline integrations/dotnet/workload-executor:17]: Superfluous section can be removed.

File: integrations/dotnet/workload-executor [hunk 8]
```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env bash
+
+set -o xtrace   # Write all commands first to stderr
```
Comment (prashantmital) [inline integrations/dotnet/workload-executor:3]: Is this really required now that things are tested and we know that this script does what it is supposed to? I think not. If that's right, you can also get rid of a lot of the `set +/-x` statements further down in this script.

File: integrations/dotnet/workload-executor [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+# Environment variables used as input:
+#   FRAMEWORK                       Set to specify .NET framework to test against. Values: "net452", "netcoreapp1.1",
+#                                   "netcoreapp2.1"
+#   ASYNC                           Whether or not to use async operations (which may be run synchronously). Values: true, false
+# Environment variables produced as output:
+#   RESULTS_DIR                     Where to output results.json
+#   ASYNC                           Whether or not to use async operations (which may be run synchronously). Values: true, false
+
+FRAMEWORK=${FRAMEWORK:-netcoreapp2.1}
+MAGIC_FILE_NAME=${MAGIC_FILE_NAME:-nox}
```
Comment (prashantmital) [inline integrations/dotnet/workload-executor:15]: This is not used anymore.

File: integrations/dotnet/workload-executor [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+else
+  RESULTS_DIR="$(pwd)"
+fi
+
+# Set a trap for the INT
+# - wait for a state change in child process bearing ID $NATIVE_WORKLOAD_EXECUTOR_PID
+# - exit with the same status/code as the exit code of the call to wait (which returns the exit code of $NATIVE_WORKLOAD_EXECUTOR_PID)
+# Note that calls to wait return immediately if the child has already changed state
+
+export RESULTS_DIR
```
Comment (prashantmital) [inline integrations/dotnet/workload-executor:45]: Is this used by the native Workload Executor?

File: integrations/dotnet/workload-executor (hunk 5 above)
Comment (MikalaiMazurenka) [reply integrations/dotnet/workload-executor]: Done.

File: integrations/dotnet/workload-executor (hunk 6 above)
Comment (MikalaiMazurenka) [reply integrations/dotnet/workload-executor]: That was done in pair with `set +x # hide connection string in test output` to hide the connection string from output log.

Removed both.

**Review by MikalaiMazurenka — COMMENTED** [review]
Ready for review

**Review by prashantmital — APPROVED** [review]
LGTM

**Pull request merged**